noir/noir-repo/aztec_macros/src/transforms/note_interface.rs

```diff
@@ -0,0 +1,510 @@
+    // Find structs annotated with #[aztec(note)]
+    let annotated_note_structs = module
+        .types
+        .iter_mut()
+        .filter(|typ| typ.attributes.iter().any(|attr| is_custom_attribute(attr, "aztec(note)")));
+
+    let mut note_fields_structs = vec![];
+
+    for note_struct in annotated_note_structs {
+        // Look for the NoteInterface trait implementation for the note
+        let trait_impl = module
+            .trait_impls
+            .iter_mut()
+            .find(|trait_impl| {
+                if let UnresolvedTypeData::Named(struct_path, _, _) = &trait_impl.object_type.typ {
+                    struct_path.last_segment() == note_struct.name
+                } else {
+                    false
+                }
+            })
+            .ok_or(AztecMacroError::CouldNotImplementNoteSerialization {
+                span: Some(note_struct.name.span()),
+                secondary_message: Some(format!(
+                    "Could not find NoteInterface trait implementation for note: {}",
+                    note_struct.name.0.contents
+                )),
+            })?;
+        // Look for the note struct implementation, generate a default one if it doesn't exist (in order to append methods to it)
+        let existing_impl = module.impls.iter_mut().find(|r#impl| match &r#impl.object_type.typ {
+            UnresolvedTypeData::Named(path, _, _) => path.last_segment().eq(&note_struct.name),
+            _ => false,
+        });
+        let note_impl = if let Some(note_impl) = existing_impl {
+            note_impl
+        } else {
+            let default_impl = TypeImpl {
+                object_type: trait_impl.object_type.clone(),
+                type_span: note_struct.name.span(),
+                generics: vec![],
+                methods: vec![],
+            };
+            module.impls.push(default_impl.clone());
+            module.impls.last_mut().unwrap()
+        };
+        // Identify the note type (struct name), its fields and its serialized length (generic param of NoteInterface trait impl)
+        let note_type = note_struct.name.0.contents.to_string();
+        let mut note_fields = vec![];
+        let note_serialized_len = match &trait_impl.trait_generics[0].typ {
+            UnresolvedTypeData::Named(path, _, _) => Ok(path.last_segment().0.contents.to_string()),
+            UnresolvedTypeData::Expression(UnresolvedTypeExpression::Constant(val, _)) => {
+                Ok(val.to_string())
+            }
+            _ => Err(AztecMacroError::CouldNotImplementNoteSerialization {
+                span: trait_impl.object_type.span,
+                secondary_message: Some(format!(
+                    "Cannot find note serialization length for: {}",
+                    note_type
+                )),
+            }),
+        }?;
+
+        // Automatically inject the header field if it's not present
```

Comment:
Should we check that the header is the correct type if the user added it?

Reply:
Switched to using the type!

Reply:
Done, it should now handle even custom headers, as long as there is exactly 1 property in the struct that has the correct type.